src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCheckedParameterTests.cs

```diff
@@ -1264,5 +1300,417 @@ .maxstack  1
+        [Fact]
+        public void TestNullCheckedInString()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(in string x!) { }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics( 
+                    // (5,30): error CS8720: By-reference parameter 'in string x!' cannot be null-checked.
+                    //     public void M(in string x!)
+                    Diagnostic(ErrorCode.ERR_NullCheckingOnByRefParameter, "!").WithArguments("in string x!").WithLocation(5, 30));
+        }
+
+        [Fact]
+        public void TestNullCheckedParams()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(params int[] number!) {}
+}";
+            CompileAndVerify(source).VerifyIL("C.M(params int[])", @"
+{
+    // Code size       15 (0xf)
+    .maxstack  1
+    IL_0000:  ldarg.1
+    IL_0001:  brtrue.s   IL_000e
+    IL_0003:  ldstr      ""number""
+    IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
+    IL_000d:  throw
+    IL_000e:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedGenericWithDefault()
+        {
+            var source = @"
+class C
+{
+    static void M1<T>(T t! = default) { }
+}";
+            CompileAndVerify(source, options: TestOptions.DebugDll).VerifyIL("C.M1<T>(T)", @"
+{
+    // Code size       21 (0x15)
+    .maxstack  1
+    IL_0000:  ldarg.0
+    IL_0001:  box        ""T""
+    IL_0006:  brtrue.s   IL_0013
+    IL_0008:  ldstr      ""t""
+    IL_000d:  newobj     ""System.ArgumentNullException..ctor(string)""
+    IL_0012:  throw
+    IL_0013:  nop
+    IL_0014:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedGenericWithDefaultClass()
+        {
+            var source = @"
+class C
+{
+    static void M2<T>(T t! = default) where T : class { }
+}";
+            CompileAndVerify(source, options: TestOptions.DebugDll).VerifyIL("C.M2<T>(T)", @"
+{
+    // Code size       21 (0x15)
+    .maxstack  1
+    IL_0000:  ldarg.0
+    IL_0001:  box        ""T""
+    IL_0006:  brtrue.s   IL_0013
+    IL_0008:  ldstr      ""t""
+    IL_000d:  newobj     ""System.ArgumentNullException..ctor(string)""
+    IL_0012:  throw
+    IL_0013:  nop
+    IL_0014:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedGenericWithDefaultStruct()
+        {
+            var source = @"
+class C
+{
+    static void M2<T>(T t! = default) where T : struct { }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                    // (4,25): error CS8718: Parameter 'T' is a non-nullable value type and cannot be null-checked.
+                    //     static void M2<T>(T t! = default) where T : struct { }
+                    Diagnostic(ErrorCode.ERR_NonNullableValueTypeIsNullChecked, "t").WithArguments("T").WithLocation(4, 25));
+        }
```

Comment:
>} [](start = 8, length = 1)

Please also test the following. I don't think we need to report `WRN_NullCheckedHasDefaultNull` for `B2<T>.F` or `B3<T>.F` (I wouldn't make any changes to try to handle these cases) but it would be good to ensure we've tested the cases:
```C#
abstract class A<T>
{
    internal abstract void F<U>(U u) where U : T;
}
class B1<T> : A<T>
{
    internal override void F<U>(U u! = default) { }
}
class B2<T> : A<T> where T : class
{
    internal override void F<U>(U u! = default) { }
}
class B3<T> : A<T?> where T : struct
{
    internal override void F<U>(U u! = default) { }
}

Reply:
Given the existing code, I'm getting no diagnostics for `B1`, `WRN_NullCheckedHasDefaultNull` for `B2`, and `ErrorCode.WRN_NullCheckingOnNullableValueType` for B3.